### ISSUE TITLE
Test Epinio on Rancher Desktop

### ISSUE THREAD

**Issue by prabalsharma** [epinio/epinio#1116]
With RD GA approaching, it will be nice to test Epinio on RD.
Will be nice to test on all platforms; linux, macos, windows. However, just testing on **linux** is acceptable.

1. Deploy RD on linux
2. Install Epinio using **helm chart**
3. some sanity test of Epinio on RD

**Comment by juadk** [comment]
Command to ssh into the lima VM:
`ssh -i ~/.local/share/rancher-desktop/lima/_config/user -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o NoHostAuthenticationForLocalhost=yes -o GSSAPIAuthentication=no -o PreferredAuthentications=publickey -o Compression=no -o BatchMode=yes -o IdentitiesOnly=yes -F /dev/null -o ControlMaster=auto -o ControlPath="~/.local/share/rancher-desktop/lima/0/ssh.sock" -o ControlPersist=5m -p 34803 127.0.0.1`

**Comment by juadk** [comment]
I'm not sure we can test epinio installation on Rancher Desktop (Linux) because of https://github.com/rancher-sandbox/rancher-desktop/issues/576

**Comment by ldevulder** [comment]
I'm playing with RD on Windows:
- installation is done without any issue
- `epinio info` can contact the epinio server
- push of an application is partially working: I have an issue during the `ANALYZING` stage, connection on `epinio-registry` is refused.

So I'm still investigating the last point.

**Comment by juadk** [comment]
I did a test on my own macbook and I was able to install and deploy an app.
However, I was forced to set s3 storage and external registry to make it work.
```
helm install epinio-installer epinio/epinio-installer --set skipTraefik=true --set useS3Storage=true --set s3Bucket=epinio-ci --set s3Endpoint=s3.amazonaws.com --set s3AccessKeyId='My access key id --set s3SecretAccessKey='My access key secret' --set externalRegistryNamespace=juadk --set externalRegistryURL=registry.hub.docker.com --set externalRegistryUsername=juadk --set externalRegistryPassword='My docker hub pw'
```
I don't know yet why it fails with minio and epinio-registry.

**Comment by evertonlperes** [comment]
Thanks for helping @juadk 

I was wondering if is possible to perform the same installation process as we had on the old school command: `epinio install  --system-domain 127.0.0.1.omg.howdoi.website` , but using helm instead.

Simulating a local development environment with Epinio and RD will be great (without passing s3 storage or even a private/public registry), similar as skipping LinkerD and Traefik, setting the parameter like, `--set s3Storage=false --set cloudRegistry=false`...

Just sharing thoughts :D

**Comment by juadk** [comment]
Of course, normally, it should work without specifying external registry and s3. Actually it works without s3 options, i just did a quick test, I'm about to update my installation log.
The minimal expected command would be `helm install epinio-installer epinio/epinio-installer --set skipTraefik=true`

**Comment by juadk** [comment]
Finally, all good on MacOS with these commands:
`helm repo add epinio https://epinio.github.io/helm-charts`
`helm install epinio-installer epinio/epinio-installer --set skipTraefik=true --set domain=192.168.1.52.omg.howdoi.website`
IP address used comes from `kubectl get svc -n kube-system traefik`

**Comment by evertonlperes** [comment]
> Finally, all good on MacOS with these commands: `helm repo add epinio https://epinio.github.io/helm-charts` `helm install epinio-installer epinio/epinio-installer --set skipTraefik=true --set domain=192.168.1.52.omg.howdoi.website` IP address used comes from `kubectl get svc -n kube-system traefik`
> 
> The issue I had was because epinio-registry doesn't work with `localhost.omg.howdoi.website`

Awesome!
I'll give a try and let you know how it goes. Thanks for helping!

**Comment by ldevulder** [comment]
So to summarize:
- RD+epinio on x86_64 MacOS is OK
- RD+epinio on x86_64 Win10 is OK (still an pure epinio issue tracked here: #1101)
- RD+epinio on x86_64 Linux is failing, tracked here: [rancher-sandbox/rancher-desktop#576](https://github.com/rancher-sandbox/rancher-desktop/issues/576)

So I think it's better to move back this issue to `Icebox` until the RD issue with port forwarding on Linux is resolved.

**Comment by evertonlperes** [comment]
On macOS the installation went well, but got `404` from the application (registry error, same as @juadk has described):

LT;DR from traefik pod:
```bash
time="2022-01-21T18:24:04Z" level=error msg="Skipping service: no endpoints found" providerName=kubernetes namespace=epinio-registry ingress=registry serviceName=registry servicePort="&ServiceBackendPort{Name:,Number:5000,}"
time="2022-01-21T18:24:06Z" level=error msg="Error configuring TLS: secret epinio-registry/epinio-registry-tls does not exist" providerName=kubernetes ingress=registry namespace=epinio-registry
time="2022-01-21T18:24:06Z" level=error msg="Skipping service: no endpoints found" serviceName=registry servicePort="&ServiceBackendPort{Name:,Number:5000,}" providerName=kubernetes ingress=registry namespace=epinio-registry
time="2022-01-21T18:24:06Z" level=error msg="Error configuring TLS: secret epinio-registry/epinio-registry-tls does not exist" namespace=epinio-registry providerName=kubernetes ingress=registry
```

[epinio-install-macOS.log](https://github.com/epinio/epinio/files/7915888/epinio-install-macOS.log)

**Comment by juadk** [comment]
@evertonlperes What did you use for domain option?
I had the same issue as you when I tried to use `localhost.omg.howdoi.website`, in the other hand, it worked when I specified the good IP address `--set domain=192.168.1.52.omg.howdoi.website`.

**Comment by evertonlperes** [comment]
> @evertonlperes What did you use for domain option? I had the same issue as you when I tried to use `localhost.omg.howdoi.website`, in the other hand, it worked when I specified the good IP address `--set domain=192.168.1.52.omg.howdoi.website`.

@juadk I've used this command:
```bash
 helm install epinio-installer epinio/epinio-installer --set skipTraefik=true --set domain=192.168.15.30.omg.howdoi.website
```

The ip address used to set the domain, I got it from `kubectl get svc -n kube-system traefik`.
output:
```bash
NAME      TYPE           CLUSTER-IP     EXTERNAL-IP     PORT(S)                      AGE
traefik   LoadBalancer   10.43.28.194   192.168.15.30   80:32647/TCP,443:30858/TCP   9s
```

Install output:
```bash
❯ helm install epinio-installer epinio/epinio-installer --set skipTraefik=true --set domain=192.168.15.30.omg.howdoi.website
NAME: epinio-installer
LAST DEPLOYED: Mon Jan 24 10:12:43 2022
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
You can get the IP address of your Ingress controller by passing this command:
`kubectl describe svc traefik --namespace traefik | awk '/Ingress/ { print $3 }'`

Make sure your domain points to the IP address of your Ingress controller (1.2.3.4).

To interact with your Epinio installation, you have to download the latest epinio binary https://github.com/epinio/epinio/releases.

Update the api location and credentials with:
`epinio config update`

For more information about Epinio, feel free to checkout https://epinio.io/ and https://docs.epinio.io/.
```

Question: Should I download and perform `epinio install` after `helm` command?

**Comment by ldevulder** [comment]
> Question: Should I download and perform `epinio install` after `helm` command?

No, only `epinio config update` is needed. Of course, epinio server and client should have the same version number.

**Comment by evertonlperes** [comment]
> > Question: Should I download and perform `epinio install` after `helm` command?
> 
> No, only `epinio config update` is needed. Of course, epinio server and client should have the same version number.

Thanks for the feedback @ldevulder 
epinio server is `v0.3.3` and my client version is `0.3.3`.
```bash
❯ epinio config update

🚢 Updating the stored credentials from the current cluster
Config: /Users/evertonperes/.config/epinio/config.yaml

✔️  Ok
```
Trying to access http://epinio.192.168.15.30.omg.howdoi.website - still getting `404`.

**Comment by juadk** [comment]
> Trying to access http://epinio.192.168.15.30.omg.howdoi.website - still getting `404`.

If I'm not wrong, this URL is not used anymore. It was used for accessing our old standalone UI. Now, our UI is Rancher.
Could you try to deploy an app and access it please?

**Comment by evertonlperes** [comment]
> > Trying to access http://epinio.192.168.15.30.omg.howdoi.website - still getting `404`.
> 
> If I'm not wrong, this URL is not used anymore. It was used for accessing our old standalone UI. Now, our UI is Rancher. Could you try to deploy an app and access it please?

I tried to deploy Rancher 2.6 along with Epinio install (helm), no luck.

Will be nice to have a chat with your guys in order to clarify some install process, still quite blurry the process on my side.

**Comment by thehejik** [comment]
This is what I achieved so far on stable `RD 1.1.1 for Linux` (installed from RPM) with `epinio cli v0.6.0`.

TL;DR
It partially works with appending the traefik https NodePort to `api` address in `settings.yaml` but it's probably unable to reach internal `epinio-registry` (also using traefik https nodeport) for some reason when doing push.

My workflow:
```
$ kubectl create ns cert-manager
$ helm upgrade --install cert-manager --namespace cert-manager jetstack/cert-manager --set installCRDs=true --set extraArgs[0]=--enable-certificate-owner-ref=true
$ helm upgrade --install epinio-installer epinio/epinio-installer --set skipTraefik=true --set skipCertManager=true --set domain=127.0.0.1.nip.io
$ nodeport=$(kubectl get svc -n kube-system traefik -o jsonpath='{.spec.ports[1].nodePort}')
$ echo $nodeport
30938
$ epinio settings update
```
Once the config is written then edit `~/.config/epinio/settings.yaml` file and append the `$nodeport` to the api address value like: `api: https://epinio.127.0.0.1.nip.io:30938` and  `epinio info` works but the push doesn't.

```
epinio --verbosity=1 --trace-level 5 push --name sample-app --path sample-app
...
Staging application with code...
2022/03/15 16:14:05 EpinioClient/Push: "level"=5 "msg"="staging code" "Name"="sample-app" "Namespace"="workspace" "Sources"="/home/thehejik/work/ecp/epinio/sample-app" "Blob"="ca8c476b-0b83-44c6-9f94-9bc5d498c3ba"
2022/03/15 16:14:05 EpinioApiClient: "level"=4 "msg"="POST https://epinio.127.0.0.1.nip.io:30938/api/v1/namespaces/workspace/applications/sample-app/stage"
2022/03/15 16:14:06 EpinioApiClient: "level"=5 "msg"="request finished"
2022/03/15 16:14:06 EpinioApiClient: "level"=5 "msg"="response received" "status"=500
2022/03/15 16:14:06 EpinioApiClient: "msg"="response is not StatusOK" "error"="Internal Server Error: no public registry URL found" "status"=500
2022/03/15 16:14:06 EpinioClient/Push: "level"=4 "msg"="return" "Name"="sample-app" "Namespace"="workspace" "Sources"="/home/thehejik/work/ecp/epinio/sample-app"
error pushing app to server: can't stage app: Internal Server Error: no public registry URL found
```
...full output [here](https://github.com/epinio/epinio/files/8255352/epinio_push_failure.log)

I'd appreciate any hint how to move forward without implementing https://github.com/rancher-sandbox/rancher-desktop/issues/576 for eg. where I can setup the `epinio-registry.127.0.0.1.nip.io:30938` value with nodeport.

I was trying to edit secret `registry-creds` where the `epinio-registry` address is stored but it didn't help.

Note: you can obtain the secret values eg. from `kubectl get secret -n workspace registry-creds -o json | jq -r '.data[".dockerconfigjson"]' | base64 -d`

**Comment by ldevulder** [comment]
> helm upgrade --install epinio-installer epinio/epinio-installer --set skipTraefik=true --set skipCertManager=true --set domain=127.0.0.1.nip.io

AFAIK `epinio-installer` shouldn't be used anymore, try to directly use `epinio` directly. I didn't tried epinio recently, I will try to do  a test this week.

**Comment by thehejik** [comment]
Thanks @ldevulder, that helped \o/

Working workflow for installing current Epinio v0.6.1 with Epinio cli v0.6.0 on RD 1.1.1 for Linux (RPM for openSUSE 15.3):
```
$ helm upgrade --install cert-manager --namespace cert-manager jetstack/cert-manager --set installCRDs=true --set extraArgs[0]=--enable-certificate-owner-ref=true --create-namespace
$ helm upgrade --install epinio --namespace epinio epinio/epinio --set skipTraefik=true --set global.domain=127.0.0.1.nip.io --create-namespace
$ wget https://github.com/epinio/epinio/releases/download/v0.6.0/epinio-linux-x86_64 -O ~/bin/epinio; chmod +x ~/bin/epinio
$ epinio settings update
```

So far it was standard installation maybe appart the `--set global.domain=127.0.0.1.nip.io` value

Now user has to modify generated `~/.config/epinio/settings.yaml` file:
```
### Find traefik ingress https nodeport
$ nodeport=$(kubectl get svc -n kube-system traefik -o jsonpath='{.spec.ports[1].nodePort}'); echo $nodeport
30910

### IMPORTANT: Append ":$nodeport" at the end of lines starting with api: and wss: strings in ~/.config/epinio/settings.yaml
$ sed -i -E "/^(wss|api):/ s/\$/:$nodeport/" ~/.config/epinio/settings.yaml

### Check communication with Epinio API
$ epinio info
✔️  Epinio Environment
Platform: k3s
Kubernetes Version: v1.22.7+k3s1
Epinio Server Version: v0.6.1
Epinio Client Version: v0.6.0
```

To deploy sample-app from https://github.com/epinio/epinio/tree/main/assets/sample-app by:
```
# First copy/clone the content of the "sample-app" folder ^
$ epinio push --name sample-app --path sample-app/
$ kubectl get ingress -n workspace
NAME                                       CLASS    HOSTS                         ADDRESS        PORTS     AGE
i-sample-app-sample-app.127.0.0.1.nip.io   <none>   sample-app.127.0.0.1.nip.io   192.168.5.15   80, 443   2m
```
Finally user can point his browser to `https://sample-app.127.0.0.1.nip.io:$nodeport` (my $nodeport has value `30910`)

**Comment by agracey** [comment]
I think `helm install epinio/epinio --set global.domain=127.0.0.1.sslip.io` works for each OS now, right? Or is it still broken on Linux? Either way, this is not our issue and we can likely close here.

**Comment by thehejik** [comment]
It's not broken on RD for Linux, there are only some limitations with port-forwarding of ports lower than 1024 on Linux, but still there are some options:
* you can access epinio API over traefik ingress `nodeport` (as described [here](https://github.com/epinio/epinio/issues/1116#issuecomment-1069103424))
* you can manually allow using of unpriviledged ports (eg 80, 443) lower than default 1024 for normal users on host by `sudo sysctl -w net.ipv4.ip_unprivileged_port_start=443` and then you can access epinio API directly on port 443

More details here https://github.com/rancher-sandbox/rancher-desktop/issues/576

This can be closed.

**Comment by juadk** [comment]
Thanks @thehejik , I think I can update our [documentation](https://docs.epinio.io/howtos/install_epinio_on_rancher_desktop) with that.
